### ISSUE TITLE
`unnecessary_safety_comment` inconsistency with wrong-style safety section

### ISSUE THREAD

**Issue by ojeda** [rust-lang/rust-clippy#15034]
### Summary

`unnecessary_safety_comment` appears to detect `/// SAFETY: ...` as a safety comment (i.e. not `# Safety` section) in a normal function, but it doesn't inside an `impl` block.

### Lint Name

unnecessary_safety_comment

### Reproducer

I tried this code:

```rust
/// SAFETY: ...
fn f() {}
```

and it triggers `unnecessary_safety_comment`. However, this:

```rust
struct S;
impl S {
    /// SAFETY: ...
    fn f() {}
}
```

does not trigger it.

I assume `unnecessary_safety_comment` is intended to only lint on actual safety comments (i.e. not safety docs/sections), even if they use `///` instead of `//`. If so, then the former is expected, and the latter is a false negative.

Please also see this other suggested lint for consistency on safety comments/docs: https://github.com/rust-lang/rust-clippy/issues/15033. For instance, if the former case above were to be a false positive, one could use that new/suggested lint to catch a case like this where the function was supposed to be `unsafe fn`. And if the former is indeed expected, then perhaps the message of the lint could be clarified if the user used `///` instead of just `//`.

More combinations: https://godbolt.org/z/43onqxYrh

### Version

```text
rustc 1.89.0-nightly (d13a431a6 2025-06-09)
binary: rustc
commit-hash: d13a431a6cc69cd65efe7c3eb7808251d6fd7a46
commit-date: 2025-06-09
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.5
```

**Comment by blyxyas** [comment]
Tip for solvers:
This happens because the compiler recognises a function from an impl as an [`ImplItem`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.ImplItem.html), while a non-associated function is identified as an [`Item`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Item.html).

[Playground Link, run Clippy on this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&code=struct+S%3B%0Aimpl+S+%7B%0A++++%23%5Bclippy%3A%3Adump%5D%0A++++%2F%2F%2F+SAFETY%3A+...%0A++++fn+f%28%29+%7B%7D%0A%7D%0A%0A%23%5Bclippy%3A%3Adump%5D%0Afn+f%28%29+%7B%7D)
